velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp

```diff
@@ -0,0 +1,388 @@
+          std::dynamic_pointer_cast<const core::ProjectNode>(plan)) {
+    return toSql(projectNode);
+  }
+  VELOX_NYI("Unsupported plan node: {}.", plan->toString());
+}
+
+std::multiset<std::vector<variant>> SparkQueryRunner::execute(
+    const std::string& sql,
+    const std::vector<RowVectorPtr>& input,
+    const RowTypePtr& resultType) {
+  return exec::test::materialize(executeVector(sql, input, resultType));
+}
+
+std::vector<RowVectorPtr> SparkQueryRunner::executeVector(
+    const std::string& sql,
+    const std::vector<RowVectorPtr>& input,
+    const RowTypePtr& resultType) {
+  auto inputType = asRowType(input[0]->type());
+  if (inputType->size() == 0) {
+    // The query doesn't need to read any columns, but it needs to see a
+    // specific number of rows. Make new 'input' as single all-null BIGINT
+    // column with as many rows as original input. This way we'll be able to
+    // create a 'tmp' table will the necessary number of rows.
+    vector_size_t numInput = 0;
+    for (const auto& v : input) {
+      numInput += v->size();
+    }
+
+    auto column = BaseVector::createNullConstant(BIGINT(), numInput, pool());
+    auto rowVector = std::make_shared<RowVector>(
+        pool(),
+        ROW({"x"}, {BIGINT()}),
+        nullptr,
+        numInput,
+        std::vector<VectorPtr>{column});
+    return executeVector(sql, {rowVector}, resultType);
+  }
+
+  // Write the input to a Parquet file.
+  auto tempFile = exec::test::TempFilePath::create();
+  const auto& filePath = tempFile->getPath();
+  auto writerPool = rootPool()->addAggregateChild("writer");
+  writeToFile(filePath, input, writerPool.get());
+
+  // Create temporary view 'tmp' in Spark by reading the generated Parquet file.
+  execute(fmt::format(
+      "CREATE OR REPLACE TEMPORARY VIEW tmp AS (SELECT * from parquet.`file://{}`);",
+      filePath));
+  return execute(sql);
+}
+
+std::vector<RowVectorPtr> SparkQueryRunner::execute(
+    const std::string& content) {
+  auto sql = google::protobuf::Arena::CreateMessage<SQL>(&arena_);
+  sql->set_query(content);
+
+  auto relation = google::protobuf::Arena::CreateMessage<Relation>(&arena_);
+  relation->set_allocated_sql(sql);
+
+  auto plan = google::protobuf::Arena::CreateMessage<Plan>(&arena_);
+  plan->set_allocated_root(relation);
+
+  auto context = google::protobuf::Arena::CreateMessage<UserContext>(&arena_);
+  context->set_user_id(kUserId);
+  context->set_user_name(kUserName);
+
+  auto request =
+      google::protobuf::Arena::CreateMessage<ExecutePlanRequest>(&arena_);
+  request->set_session_id(kSessionId);
+  request->set_allocated_user_context(context);
+  request->set_allocated_plan(plan);
+
+  grpc::ClientContext clientContext;
+  auto reader = stub_->ExecutePlan(&clientContext, *request);
+  ExecutePlanResponse response;
+  std::vector<RowVectorPtr> results;
+  while (reader->Read(&response)) {
+    if (response.session_id() == kSessionId && response.has_arrow_batch()) {
```

Comment:
It could be the case that several clients are submitting requests to a spark connect server at the same time, and session id could be used to make sure expected results are fetched. But in the use case of Velox fuzzer test, we only submit from one client.

Reply:
Interesting. Are you saying it is possible that a client may receive a response for request made from another client? Would that be a bug? Should this check be an assert?

Reply:
Thanks for noticing. I will change it as an assert.

The grpc server should handle each client's requests independently and sends the responses back on the same stream. I agree it should be a bug if the responses of different clients are mixed.

In the case where a single client has multiple active sessions, the sessionId could be useful to help fetch the responses of specific session. But the scenario of Velox fuzzer test is not that case.